src/optimagic/optimizers/pygad_optimizer.py

```diff
@@ -0,0 +1,292 @@
+            )
+
+            initial_population[0] = x0
+
+        gene_space = [
+            {"low": problem.bounds.lower[i], "high": problem.bounds.upper[i]}
+            for i in range(len(x0))
+        ]
+
+        ga_instance = pygad.GA(
+            num_generations=self.num_generations,
+            num_parents_mating=num_parents_mating,
+            fitness_func=fitness_function,
+            fitness_batch_size=effective_fitness_batch_size,
+            initial_population=initial_population,
+            gene_space=gene_space,
+            parent_selection_type=self.parent_selection_type,
+            keep_parents=self.keep_parents,
+            keep_elitism=self.keep_elitism,
+            K_tournament=self.K_tournament,
+            crossover_type=self.crossover_type,
+            crossover_probability=self.crossover_probability,
+            mutation_type=self.mutation_type,
+            mutation_probability=self.mutation_probability,
+            mutation_by_replacement=self.mutation_by_replacement,
+            mutation_percent_genes=self.mutation_percent_genes,
+            mutation_num_genes=self.mutation_num_genes,
+            random_mutation_min_val=self.random_mutation_min_val,
+            random_mutation_max_val=self.random_mutation_max_val,
+            allow_duplicate_genes=self.allow_duplicate_genes,
+            gene_constraint=self.gene_constraint,
+            sample_size=self.sample_size,
+            stop_criteria=self.stop_criteria,
+            parallel_processing=None,
+            random_seed=self.random_seed,
+        )
+
+        ga_instance.run()
+
+        res = _process_pygad_result(ga_instance)
+
+        return res
+
+
+def determine_effective_batch_size(
+    fitness_batch_size: int | None, n_cores: int
+) -> int | None:
+    """Determine the effective fitness_batch_size for parallel processing.
+
+    Behavior:
+    - If `fitness_batch_size` is explicitly provided:
+        - The value is returned unchanged.
+        - A warning is issued if it is less than `n_cores`, as this may
+        underutilize available cores.
+    - If `fitness_batch_size` is `None`:
+        - If `n_cores` > 1, defaults to `n_cores`.
+        - Otherwise, returns None (i.e., single-threaded evaluation).
+    Args:
+        fitness_batch_size: User-specified batch size or None
+        n_cores: Number of cores for parallel processing
+
+    Returns:
+        Effective batch size for PyGAD, or None for single-threaded processing
+
+    """
+    if fitness_batch_size is not None:
+        if fitness_batch_size < n_cores:
+            warnings.warn(
+                f"fitness_batch_size ({fitness_batch_size}) is smaller than "
+                f"n_cores ({n_cores}). This may reduce parallel efficiency. "
+                f"Consider setting fitness_batch_size >= n_cores."
+            )
+        return fitness_batch_size
+    elif n_cores > 1:
+        return n_cores
+    else:
+        return None
```

Comment:
Please use one return statement at the end. I.e. instead of returning multiple times, assign to a variable called `out` or `result` and then return that at the end.